File: tool/create_food_db.dart
// Script tạo food_database.db trực tiếp bằng sqlite3
// Chạy: dart run tool/create_food_db.dart

import 'dart:io';
import 'dart:convert';
import 'package:sqlite3/sqlite3.dart';

void main() {
  // Đọc JSON đã generate
  final jsonFile = File('assets/database/food_database.json');
  final foods = jsonDecode(jsonFile.readAsStringSync()) as List<dynamic>;
  print('Loaded ${foods.length} food items from JSON');

  // Tạo database
  final dbPath = 'assets/database/food_database.db';
  final dbFile = File(dbPath);
  if (dbFile.existsSync()) {
    dbFile.deleteSync();
    print('Deleted existing database');
  }

  final db = sqlite3.open(dbPath);

  // Tạo bảng
  db.execute('''
    CREATE TABLE IF NOT EXISTS foods (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      key TEXT NOT NULL UNIQUE,
      name TEXT NOT NULL,
      name_en TEXT NOT NULL,
      calories_per_100g REAL NOT NULL,
      glycemic_index INTEGER NOT NULL,
      protein REAL NOT NULL,
      carbs REAL NOT NULL,
      fat REAL NOT NULL,
      fiber REAL NOT NULL,
      category TEXT NOT NULL
    )
  ''');

  // Tạo indexes
  db.execute('CREATE INDEX idx_foods_name ON foods(name)');
  db.execute('CREATE INDEX idx_foods_name_en ON foods(name_en)');
  db.execute('CREATE INDEX idx_foods_category ON foods(category)');
  db.execute('CREATE INDEX idx_foods_key ON foods(key)');

  // Insert dữ liệu
  final stmt = db.prepare('''
    INSERT OR REPLACE INTO foods (key, name, name_en, calories_per_100g, glycemic_index, protein, carbs, fat, fiber, category)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  ''');

  db.execute('BEGIN TRANSACTION');
  for (final food in foods) {
    stmt.execute([
      food['key'],
      food['name'],
      food['name_en'],
      food['calories_per_100g'],
      food['glycemic_index'],
      food['protein'],
      food['carbs'],
      food['fat'],
      food['fiber'],
      food['category'],
    ]);
  }
  db.execute('COMMIT');
  stmt.dispose();

  // Verify
  final result = db.select('SELECT COUNT(*) as count FROM foods');
  print('Database created: $dbPath');
  print('Total records: ${result.first['count']}');

  // Liệt kê categories
  final categories =
      db.select('SELECT category, COUNT(*) as count FROM foods GROUP BY category ORDER BY count DESC');
  print('\nCategories (${categories.length}):');
  for (final row in categories) {
    print('  - ${row['category']} (${row['count']} items)');
  }

  db.dispose();
  print('\n✅ food_database.db created successfully!');
}

File: tool/create_food_db.py
"""
Tạo lại food_database.db từ food_database.json
Chạy: python tool/create_food_db.py
"""

import json
import sqlite3
import os

def main():
    json_path = 'assets/database/food_database.json'
    db_path = 'assets/database/food_database.db'

    # Load JSON
    with open(json_path, 'r', encoding='utf-8') as f:
        foods = json.load(f)
    print(f"Loaded {len(foods)} food items from JSON")

    # Xóa DB cũ
    if os.path.exists(db_path):
        os.remove(db_path)
        print("Deleted existing database")

    # Tạo DB mới
    conn = sqlite3.connect(db_path)
    cur = conn.cursor()

    # Tạo bảng
    cur.execute('''
        CREATE TABLE IF NOT EXISTS foods (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            key TEXT NOT NULL UNIQUE,
            name TEXT NOT NULL,
            name_en TEXT NOT NULL,
            calories_per_100g REAL NOT NULL,
            glycemic_index INTEGER NOT NULL,
            protein REAL NOT NULL,
            carbs REAL NOT NULL,
            fat REAL NOT NULL,
            fiber REAL NOT NULL,
            category TEXT NOT NULL
        )
    ''')

    # Tạo indexes
    cur.execute('CREATE INDEX idx_foods_name ON foods(name)')
    cur.execute('CREATE INDEX idx_foods_name_en ON foods(name_en)')
    cur.execute('CREATE INDEX idx_foods_category ON foods(category)')
    cur.execute('CREATE INDEX idx_foods_key ON foods(key)')

    # Insert dữ liệu
    for food in foods:
        cur.execute('''
            INSERT OR REPLACE INTO foods (key, name, name_en, calories_per_100g, glycemic_index, protein, carbs, fat, fiber, category)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ''', (
            food['key'],
            food['name'],
            food.get('name_en', ''),
            food.get('calories_per_100g', 0),
            food.get('glycemic_index', 50),
            food.get('protein', 0),
            food.get('carbs', 0),
            food.get('fat', 0),
            food.get('fiber', 0),
            food.get('category', 'Chưa phân loại'),
        ))

    conn.commit()

    # Verify
    cur.execute('SELECT COUNT(*) FROM foods')
    count = cur.fetchone()[0]
    print(f"\nDatabase created: {db_path}")
    print(f"Total records: {count}")

    # Liệt kê categories
    cur.execute('SELECT category, COUNT(*) as cnt FROM foods GROUP BY category ORDER BY cnt DESC')
    categories = cur.fetchall()
    print(f"\nCategories ({len(categories)}):")
    for cat, cnt in categories:
        print(f"  - {cat} ({cnt} items)")

    conn.close()
    print(f"\n✅ food_database.db created successfully!")

if __name__ == '__main__':
    main()
